pkg/convert/dynamic.go

```diff
@@ -0,0 +1,185 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convert
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// To support DynamicPseudoType, perform a best-effort conversion at the data level without any type information.
+func newDynamicEncoder() Encoder {
+	return &dynamicEncoder{}
+}
+
+type dynamicEncoder struct{}
+
+// Convert a PropertyValue to a tftypes.Value. If the optional ty parameter is specified, ensure that the resulting
+// tftypes.Value is constructed with the given type.
+func (enc *dynamicEncoder) fromPropertyValue(p resource.PropertyValue) (tftypes.Value, error) {
+	switch {
+	case propertyValueIsUnknown(p):
+		return tftypes.NewValue(tftypes.Object{}, tftypes.UnknownValue), nil
+	case p.IsNull():
+		return tftypes.NewValue(tftypes.Object{}, nil), nil
+	case p.IsBool():
+		return tftypes.NewValue(tftypes.Bool, p.BoolValue()), nil
+	case p.IsNumber():
+		return tftypes.NewValue(tftypes.Number, p.NumberValue()), nil
+	case p.IsString():
+		return tftypes.NewValue(tftypes.String, p.StringValue()), nil
+	case p.IsArray():
+		var result []tftypes.Value
+		for _, e := range p.ArrayValue() {
+			te, err := enc.fromPropertyValue(e)
+			if err != nil {
+				return tftypes.Value{}, err
+			}
+			result = append(result, te)
+		}
+		// tftypes.NewValue is pretty strict in that array elements must have the same type.
+		// Try to encode this as an array and fallback on a tuple if the types do not match.
+		if commonTy, err := tftypes.TypeFromElements(result); err == nil {
+			return tftypes.NewValue(tftypes.List{ElementType: commonTy}, result), nil
+		} else {
+			types := []tftypes.Type{}
+			for _, r := range result {
+				types = append(types, r.Type())
+			}
+			return tftypes.NewValue(tftypes.Tuple{ElementTypes: types}, result), nil
+		}
+	case p.IsAsset():
+		return tftypes.Value{}, fmt.Errorf("Assets inside dynamically typed blocks are not yet supported")
+	case p.IsArchive():
+		return tftypes.Value{}, fmt.Errorf("Archives inside dynamically typed blocks are not yet supported")
+	case p.IsSecret() || p.IsOutput() && p.OutputValue().Secret:
+		return tftypes.Value{}, fmt.Errorf("Secrets inside dynamically typed blocks are not yet supported")
+	case p.IsOutput():
+		o := p.OutputValue()
+		contract.Assertf(o.Known && !o.Secret,
+		  "Cannot be unknown at this point, or secret.")
+		return enc.fromPropertyValue(p.OutputValue().Element)
+	case p.IsObject():
+		// Maps and objects are confused in this Pulumi representation, we cannot reliably tell them apart.
```

Comment:
Does the handling between object/map depend on individual providers to implement?

Reply:
I'm not 100% sure what you're suggesting here. The provider is mostly out of the loop. If the user specifies `{"A": "B"}` we need to decide if this is a map or an object, because in the Pulumi representation both are the same, but in tftypes representation they're different things. We can refine the guess if we run into trouble with the current one.

Reply:
I'm wondering how it'd work in TF - when the provider author specifies the type to be dynamic, would they expect users to specify the parameter as an object or map? Would they handle both in the TF provider?

Reply:
I'm expecting that if the provider deserializes into typed models this distinction is made moot. I added a few test cases on this (homogeneous and heterogeneous maps). LMK if you think this needs another test case to cover something of concern.